### PR TITLE
add Laravel ServiceProvider file and configuration

### PR DESCRIPTION
**Pull request by sandrokeil**
There are some problems with merging of configuration, because Laravel doesn't use `array_merge_recursive()` [here](https://github.com/laravel/framework/blob/5.2/src/Illuminate/Support/ServiceProvider.php#L67). For example the event_bus plugins are removed after merge.

A workaround is to have the full configuration in the user config files, but documentation says 

_You may also choose to merge your own package configuration file with the application's copy. This allows your users to include only the options they actually want to override in the published copy of the configuration_.

We provide a default package configuration and the user have to add some configuration to some sub keys. See [README.md](https://github.com/sandrokeil/prooph-package#example) for more details.

### CODE REVIEW

**Comment by bweston92** [comment]
@sandrokeil normally one would "publish" the config therefore all the defaults are there and they just update the defaults making a recursive merge not required.

**Comment by sandrokeil** [comment]
Ok, that's a Laravel feature so I have to update the readme and it so it's the normal Laravel App behaviour. Thank's.

**Comment by bweston92** [comment]
Or you can use my PR to your repository? If not close it.

**Comment by sandrokeil** [comment]
@bweston92 Is this [readme](https://github.com/sandrokeil/prooph-package/blob/master/README.md) understandable? Then we can merge it.

**Comment by bweston92** [comment]
@sandrokeil I just don't like how 2 service providers are required. I know it is a must to support container-interop, hmm. Other then that it is fine :+1:

**Comment by sandrokeil** [comment]
@bweston92 Thanks for the response. The main reason is, that we can use the original factories with config validation. Maybe Laravel does support container-interop in the future. :-)

**Comment by bweston92** [comment]
@taylorotwell thoughts on Laravel with container-interop support without third-party package?

**Comment by taylorotwell** [comment]
I don't have a big problem with it if someone wants to send over a PR.

File: src/ProophServiceProvider.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            \Prooph\EventStore\Snapshot\SnapshotStore::class,
+            // service bus
+            \Prooph\ServiceBus\CommandBus::class,
+            \Prooph\ServiceBus\EventBus::class,
+            \Prooph\EventStore\Adapter\Doctrine\DoctrineEventStoreAdapter::class,
+            \Prooph\EventStore\Snapshot\Adapter\Doctrine\DoctrineSnapshotAdapter::class,
+            // prooph/event-store-bus-bridge set up
+            \Prooph\EventStoreBusBridge\TransactionManager::class,
+            \Prooph\EventStoreBusBridge\EventPublisher::class,
+            'Prooph\\EventStore\\Adapter\\Doctrine\\DoctrineEventStoreAdapter',
```
Comment (bweston92) [inline src/ProophServiceProvider.php]: @sandrokeil what's wrong with `Prooph\EventStore\Adapter\Doctrine\DoctrineEventStoreAdapter::class` ?

Comment (bweston92) [inline src/ProophServiceProvider.php]: It won't attempt to load class mainly for performance reasons. It's just a constant: https://3v4l.org/qFEep

Comment (sandrokeil) [inline src/ProophServiceProvider.php]: @bweston92 Good catch! I will fix it. I will fix it for dependencies too.

**Comment by sandrokeil** [comment]
- Updated the code to use FCQN instead of strings for service names
- Moved `proophsoftware/prooph-cli` to suggest, it's a development tool
- Use only factories for services
- Supports `Zend\ServiceManager` v2 and v3

**Comment by sandrokeil** [comment]
@bweston92 Are you working on a Laravel PR for container-interop?

**Comment by bweston92** [comment]
@sandrokeil not got time this week, busy week - sorry

File: config/dependencies.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<?php
+/**
+ * prooph (http://getprooph.org/)
+ *
+ * @see       https://github.com/proophsoftware/prooph-package for the canonical source repository
+ * @copyright Copyright (c) 2016 prooph software GmbH (http://prooph-software.com/)
+ * @license   https://github.com/proophsoftware/prooph-package/blob/master/LICENSE.md New BSD License
+ */
+
+use Zend\ServiceManager\Factory\InvokableFactory;
```
Comment (codeliner) [inline config/dependencies.php]: Is this the only reason for the zend/service-manager dependency? If so we should provide our own InvokableFactory.

**Comment by sandrokeil** [comment]
@codeliner I've updated the dependencies and added our own factory. This is a better solution. :+1:

**Comment by codeliner** [comment]
thanks for the great work @sandrokeil and @bweston92 for review

**Pull request merged**